### DISCUSSION TITLE
File Upload to multiple S3 buckets

### DISCUSSION THREAD

**Discussion by ihsan-96** [directus/directus#9995] [Archive]
Hi Team,

I have configured multiple STORAGE_LOCATIONS, which includes two S3 buckets one accessed by public and another accessed only through directus.

the config looks like this

> STORAGE_LOCATIONS="aws,public,local"
> 
> STORAGE_LOCAL_DRIVER="local"
> STORAGE_LOCAL_ROOT="./uploads"
> 
> STORAGE_AWS_DRIVER="s3"
> STORAGE_AWS_KEY="xxxxxxxx"
> STORAGE_AWS_SECRET="xxxxxxxx"
> STORAGE_AWS_BUCKET="bucket1"
> STORAGE_AWS_REGION="xxxxxxxx"
> STORAGE_AWS_ENDPOINT="s3.xxxxxxxx.amazonaws.com"
> 
> STORAGE_PUBLIC_DRIVER="s3"
> STORAGE_PUBLIC_KEY="xxxxxxxx"
> STORAGE_PUBLIC_SECRET="xxxxxxxx"
> STORAGE_PUBLIC_BUCKET="bucket2"
> STORAGE_PUBLIC_REGION="xxxxxxxx"
> STORAGE_PUBLIC_ENDPOINT="s3.xxxxxxxx.amazonaws.com"

The [Documentation](https://docs.directus.io/configuration/config-options/#file-storage) says,
`You can also configure multiple storage adapters at the same time. This allows you to choose where files are being uploaded on a file-by-file basis. In the Admin App, files will automatically be uploaded to the first configured storage location`

this feels contradictory to me as it is automatically stored to the first location.
I am not able to find how we can select different storage based on files.

Is there a way how we can configure where to store, based on file directory / collection / or any other value.

**Comment by rijkvanzanten** [comment]
> this feels contradictory to me as it is automatically stored to the first location.

How so? That's exactly what it says:

"In the Admin App, files will automatically be uploaded to the first configured storage location"

The API currently supports uploading/download files from any storage adapter, and the app will show the files regardless of storage adapter used. That being said, the app will currently only use the first configured storage adapter when uploading new files from the app.

**Reply by ihsan-96** [reply]
Ok, So using API would allow uploading to multiple storage adapter, but via admin app it's not possible.

Please consider adding this feature as it is necessary for building scalable systems. Even though building a custom interface using the API is a solution, it takes away the simplicity of having a admin app.

**Comment by ShivamBh** [comment]
If I have two fields "video_src" and "thumb_src" respectively on a model, and I have configured different STORAGE_LOCATIONS such as the one above, how can I get the fields to upload to the bucket of choice? 

Looking through the docs, one solution I can think of is making my own **upload interface** which can read config values directly(maybe even set a default), but that is a bit involved. I am looking for a solution using the existing API, maybe using hooks. Of course if such a solution does not exist with the "out-of-the-box" tools provided, I will go ahead with the interface one. But I get the feeling I might be overlooking something in the docs, in which case maybe you can point me in the right direction.

**Reply by rijkvanzanten** [reply]
> If I have two fields "video_src" and "thumb_src" respectively on a model, and I have configured different STORAGE_LOCATIONS such as the one above, how can I get the fields to upload to the bucket of choice?

There's no way to do that from the app at this point. Making a custom interface would be the only way.

**Reply by ShivamBh** [reply]
Thank you :)

**Comment by dirkdirk** [comment]
I'm confused by the docs also.

Statement 1
This allows you to choose where files are being uploaded on a file-by-file basis.

Statement 2
In the Admin App, files will automatically be uploaded to the first configured storage location.

If I'm allowed to choose, then why are the files only uploaded to the first configured location?

Clearly I do not understand the context of the first or second statement.  Can someone please explain?  Thanks!

**Reply by Dominic-Marcelino** [reply]
As Rjik mentioned above you can choose the storage adapter when using the api. When using the App it will always use the first given provider

**Reply by dirkdirk** [reply]
Got it!

Statement 1 applies to the API.

Statement 2 applies to the Admin App.
